### PR TITLE
Make all the OpenStack resources be managed by a Heat Stack

### PR DESCRIPTION
**Pull request by lhuard1A**
The goal of this change is to have all the OpenStack resources be managed by a single Heat Stack.

Previously, only the network infrastructure was handled by a Heat Stack. All the VMs and their associated floating IPs were managed “one by one” directly from Ansible.

The advantage of the single Heat Stack approach are:

## Rapidity

Ansible is able to run in parallel tasks done on different hosts. But before that PR, the VMs were spawned from a `with_items` loop. All `with_items` iterations were sequentially executed on the single `localhost` host.
By using a Heat template for spawning the VMs, all the VMs can be spawned in parallel and that parallelism is on Heat engine side.

#### Timing with current `master`
On the OpenStack instance I have access to, with the current `master` code, creating a 5 nodes cluster took:
```
real    11m3.698s
user    0m21.987s
sys     0m6.903s
```
The top most consuming tasks were:
```
Launch instance(s) ---------------------------------------------------- 126.71s
openshift_node | Allow NFS access for VMs ------------------------------ 69.85s
openshift_node | Start and enable openshift-node ----------------------- 27.90s
openshift_facts  ------------------------------------------------------- 27.77s
Wait for ssh ----------------------------------------------------------- 27.44s
openshift_node | Set node OpenShift facts ------------------------------ 24.35s
openshift_facts | Gather OpenShift facts ------------------------------- 16.71s
openshift_master | Create the master certificates if they do not already exist -- 14.80s
Evaluate oo_hosts_to_update -------------------------------------------- 14.57s
openshift_common | Set common OpenShift facts -------------------------- 12.66s
```

#### Timing with that PR
With that PR, creating a 5 nodes cluster takes:
```
real    7m57.965s
user    0m26.563s
sys     0m7.517s
```
The top most consuming tasks are now:
```
openshift_node | Allow NFS access for VMs ------------------------------ 65.05s
Wait for OpenStack Stack readiness ------------------------------------- 59.15s
openshift_facts  ------------------------------------------------------- 29.21s
openshift_node | Set node OpenShift facts ------------------------------ 22.48s
openshift_node | Start and enable openshift-node ----------------------- 22.42s
Wait for ssh ----------------------------------------------------------- 20.11s
openshift_facts | Gather OpenShift facts ------------------------------- 17.78s
openshift_master | Create the master certificates if they do not already exist -- 16.18s
Evaluate oo_hosts_to_update -------------------------------------------- 14.24s
openshift_common | Set common OpenShift facts -------------------------- 14.11s
```

### cluster termination
The cluster termination is greatly simplified as well since is only consists in deleting the Heat stack.

#### Timing with current `master`
5 nodes cluster termination with current `master`:
```
real    1m56.375s
user    0m7.323s
sys     0m1.147s
```
Top consuming tasks:
```
Terminate instance(s) -------------------------------------------------- 90.49s
add_host  -------------------------------------------------------------- 10.44s
Delete floating IPs ----------------------------------------------------- 5.79s
Retrieve the floating IPs ----------------------------------------------- 5.05s
Destroy the network ----------------------------------------------------- 1.47s
set_fact cluster_group=tag_env_{{ cluster_id }} ------------------------- 0.01s
```

#### Timing with that PR
```
real    1m25.369s
user    0m16.413s
sys     0m2.203s
```
Top consuming tasks:
```
Wait for the completion of the OpenStack Stack deletion ---------------- 81.55s
Delete the OpenStack Stack ---------------------------------------------- 1.21s
```

## Avoid IaaS resources leaks

Whatever happens, all the resources associated to a given cluster can be freed by deleting the Heat stack, even if the user aborts the `openshift-ansible` script at any time.

## Customization

OpenStack deployment configuration is done at a single place.
Let’s imagine that I want to attach a volume on every machine — to be used as a docker backend storage. I need to create one per VM, to attach them to VMs, and to take care of detaching and deleting them at cluster termination.
With that PR, everything will be done inside `heat_stack_server.yaml` and the cluster termination code is not impacted at all.

### CODE REVIEW

File: playbooks/openstack/openshift-cluster/terminate.yml [hunk 1]
```diff
@@ -5,39 +5,18 @@
...
-    with_items: groups['oo_hosts_to_terminate'] | default([])
-
-  - name: Delete floating IPs
-    command: "neutron floatingip-delete {{ item.stdout }}"
-    with_items: floating_ips_to_delete.results | default([])
-
-  - name: Destroy the network
-    command: "heat stack-delete {{ openstack_network_prefix }}-stack"
+  - name: Delete the OpenStack Stack
+    command: 'heat stack-delete openshift-ansible-{{ cluster_id }}-stack'
```
Comment (akram) [inline playbooks/openstack/openshift-cluster/terminate.yml]: @lhuard1A ,what do you think about unsubscribing from RHN here in case of a RHEL stack ?

Comment (lhuard1A) [inline playbooks/openstack/openshift-cluster/terminate.yml]: Hi @akram,
I think we should unsubscribe from RHN if we have subscribed during the cluster creation. And both might be conditioned by the `enterprise` deployment type.
Anyhow, I dealt with RHN subscription in a separate PR #264.
Could you review this one?

**Comment by lhuard1A** [comment]
Hi @twiest , @detiber . Could you please review this PR? The idea is to manage all the IaaS resources under a single “Heat stack”

**Comment by twiest** [comment]
@lhuard1A Looks good to me.

@detiber Let me know what you think. If you're a +1, I'll merge this.

**Comment by detiber** [comment]
@twiest +1 from me

**Pull request merged**